### DISCUSSION TITLE
Custom Atom Types and Features in Reaction SMARTS

### DISCUSSION THREAD

**Discussion by t-raymond** [rdkit/rdkit#7293] [Q&A]
Hello!

Is there a way to use the atom types and features specified in the BaseFeatures.fdef file when defining SMARTS reaction templates?

As a simple example, if I were to define a deprotonation reaction with `AllChem.ReactionFromSmarts()`, I would like to be able to use the AcidicGroup feature that is defined in the fdef file as:

```
DefineFeature AcidicGroup [C,S](=[O,S,P])-[O;H1,H0&-1]
  Family NegIonizable
  Weights 1.0,1.0,1.0
EndFeature
```

I have tried a number of different ways to write out the SMARTS string, but always get the same error related to parsing the string

`rxn = AllChem.ReactionFromSmarts('[AcidicGroup:1]>>[*-:1]')`
`ValueError: ChemicalReactionParserException: Problems constructing reactant from SMARTS: [AcidicGroup:1]`

I'm quite new to using RDKit, so I'm sure there are multiple things I may be doing incorrectly. 

My project will involve defining a lot of reaction templates, so being able to define and use custom features would be exceptionally helpful.

Thank you all in advance!

**Comment by bp-kelley** [comment]
We do have a way of doing this but it's a bit tricky.  It uses the PreProcessReaction to look for atom labels that can be used to attach more specific functional group smarts patterns.

```
help(rdChemReactions.PreprocessReaction)
Help on built-in function PreprocessReaction in module rdkit.Chem.rdChemReactions:

PreprocessReaction(...)
    PreprocessReaction( (ChemicalReaction)reaction [, (dict)queries={} [, (str)propName='molFileValue']]) -> object :
        A function for preprocessing reactions with more specific queries.
        Queries are indicated by labels on atoms (molFileAlias property by default)
        When these labels are found, more specific queries are placed on the atoms.
        By default, the available quieries come from 
          FilterCatalog.GetFlattenedFunctionalGroupHierarchy(True)n
        Sample Usage:
          >>> from rdkit import Chem, RDConfig
          >>> from rdkit.Chem import MolFromSmiles, AllChem
          >>> from rdkit.Chem.rdChemReactions import PreprocessReaction
          >>> import os
          >>> testFile = os.path.join(RDConfig.RDCodeDir,'Chem','SimpleEnum','test_data','boronic1.rxn')
          >>> rxn = AllChem.ReactionFromRxnFile(testFile)
          >>> rxn.Initialize()
          >>> nWarn,nError,nReacts,nProds,reactantLabels = PreprocessReaction(rxn)
```

It may be a tad tricky to set up but, here is an example making Bromine match only aromatic bromines:

```
>>> rxn = rdChemReactions.ChemicalReaction()
>>> q = Chem.MolFromSmarts("Br[*:1]")
>>> q.GetAtomWithIdx(0).SetProp("functional_group", "halogen.bromine.aromatic")
>>> p = Chem.MolFromSmarts("[*:1]")
>>> rxn.AddReactantTemplate(q)
1
>>> rxn.AddProductTemplate(p)
1
>>> rdChemReactions.ReactionToSmarts(rxn)
'Br[*:1]>>[*:1]'
>>> rdChemReactions.PreprocessReaction(rxn, propName='functional_group')
(0, 0, 1, 1, (((0, 'halogen.bromine.aromatic'),),))
>>> rdChemReactions.ReactionToSmarts(rxn)
'[Br&$([Br&$(Br-&!@c)])][*:1]>>[*:1]'
```

The full list of functional groups can be found:

```
>>> from rdkit.Chem import FilterCatalog
>>> FilterCatalog.GetFlattenedFunctionalGroupHierarchy(True)
```
